oci/pull.bzl

```diff
@@ -0,0 +1,385 @@
+        "blobs",
+        "oci-layout",
+        "index.json",
+    ],
+)
+"""
+
+def _trim_hash_algorithm(digest):
+    parts = digest.split(":", 1)
+    if len(parts) != 2:
+        fail("digest {} doesn't contain a colon".format(digest))
+    return parts[1]
+
+def _find_platform_manifest(rctx, image_mf):
+    for mf in image_mf["manifests"]:
+        plat = "{}/{}".format(mf["platform"]["os"], mf["platform"]["architecture"])
+        if plat == rctx.attr.platform:
+            return mf
+    fail("No matching manifest found in image {} for platform {}".format(rctx.attr.image, rctx.attr.platform))
+
+def _oci_pull_impl(rctx):
+    mf_file = _trim_hash_algorithm(rctx.attr.digest)
+    mf, mf_len = _download(rctx, rctx.attr.digest, mf_file, resource = "manifests")
+
+    if mf["mediaType"] == _MANIFEST_TYPE:
+        if rctx.attr.platform:
+            fail("{} is a single-architecture image, so attribute 'platform' should not be set.")
+        image_mf_file = mf_file
+        image_mf = mf
+        image_mf_len = mf_len
+        image_digest = rctx.attr.digest
+    elif mf["mediaType"] == _MANIFEST_LIST_TYPE:
+        # extra download to get the manifest for the selected arch
+        if not rctx.attr.platform:
+            fail("{} is a multi-architecture image, so attribute 'platform' is required.")
+        matching_mf = _find_platform_manifest(rctx, mf)
+        image_digest = matching_mf["digest"]
+        image_mf_file = _trim_hash_algorithm(image_digest)
+        image_mf, image_mf_len = _download(rctx, image_digest, image_mf_file, resource = "manifests")
+    else:
+        fail("Unrecognized mediaType {} in manifest file".format(image_mf["mediaType"]))
+
+    image_config_file = _trim_hash_algorithm(image_mf["config"]["digest"])
+    image_config = _download(rctx, image_mf["config"]["digest"], image_config_file)
+    tars = []
+    for layer in image_mf["layers"]:
+        hash = _trim_hash_algorithm(layer["digest"])
+
+        # TODO: we should avoid eager-download of the layers ("shallow pull")
+        _download(rctx, layer["digest"], hash)
+        tars.append(hash)
+
+    # To make testing against `crane pull` simple, we take care to produce a byte-for-byte-identical
+    # index.json file, which means we can't use jq (it produces a trailing newline) or starlark
+    # json.encode_indent (it re-orders keys in the dictionary).
+    if rctx.attr.platform:
+        os, arch = rctx.attr.platform.split("/", 1)
+        index_mf = """\
+{
+   "schemaVersion": 2,
+   "manifests": [
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
+         "size": %s,
+         "digest": "%s",
+         "platform": {
+            "architecture": "%s",
+            "os": "%s"
+         }
+      }
+   ]
+}""" % (image_mf_len, image_digest, arch, os)
+    else:
+        index_mf = """\
+{
+   "schemaVersion": 2,
+   "manifests": [
+      {
+         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
```

Comment:
this media type would be wrong the pulled image has oci media types rather than docker v2. the media type of the remote manifest should determine this. `application/vnd.oci.image.manifest.v1+json` or `application/vnd.docker.distribution.manifest.v2+json` depending on the mediaType of the image.

Reply:
eg: `crane pull thesayyn/debian:oci ./pco --format=oci`

Reply:
yeah https://github.com/bazel-contrib/rules_oci/issues/74 will catch this too

Reply:
great.